### PR TITLE
Add build rule for iOS static frameworks

### PR DESCRIPTION
**Pull request by rajivshah3**
See https://gitlab.com/iota-foundation/software/entangled/merge_requests/41

# Test Plan:
Tests can be run with:
```
xcodebuild test -project mobile/ios/Demo/Demo.xcodeproj -scheme EntangledDemo -destination 'platform=iOS Simulator,name=iPhone 6s'
```

### CODE REVIEW

File: mobile/ios/Interface.m [hunk 1]
```diff
@@ -0,0 +1,27 @@
+//
+//  Interface.m
+//  
+//
+//  Created by Rajiv Shah on 4/13/18.
+//
```
Comment (jserv) [inline mobile/ios/Interface.m]: Should append copyright information.

Comment (rajivshah3) [reply mobile/ios/Interface.m]: Good point. Will do 👍

File: .gitignore [hunk 2]
```diff
@@ -1,76 +1,149 @@
...
-bazel*
-.bazelrc
-cclient/bazel*
-cclient/cmake-*
-
-# Atom conf
-.clang_complete
+*.class
+
+# Mobile Tools for Java (J2ME)
```
Comment (th0br0) [inline .gitignore]: Please remove these useless gitignores.

File: README.md [hunk 3]
```diff
@@ -30,3 +30,37 @@ docker cp ##path_to_configuration_file ##the_name:/app/default_configuration.yam
 ```shell
 docker start ##the_name
 ```
+
+## iOS
```
Comment (th0br0) [inline README.md]: This should go into the relevant subdirectory, not the root README

File: WORKSPACE [hunk 4]
```diff
@@ -16,6 +16,18 @@ android_ndk_repository(
     api_level = 14,
 )
 
+git_repository(
```
Comment (th0br0) [inline WORKSPACE:40]: These dependencies are outdated, please update

File: common/helpers/digest.c [hunk 5]
```diff
@@ -4,6 +4,9 @@
 #include "common/trinary/trit_tryte.h"
 #include "digest.h"
 
+#include <string.h>
```
Comment (th0br0) [inline common/helpers/digest.c]: I believe a recent commit by Thibault conflicts with this

File: mobile/ios/BUILD [hunk 6]
```diff
@@ -0,0 +1,16 @@
+load("@build_bazel_rules_apple//apple:ios.bzl", "ios_static_framework")
+
+ios_static_framework(
+    name="ios_bindings",
```
Comment (th0br0) [inline mobile/ios/BUILD]: I think you need to run `buildifier` here - please make sure that you've got the hooks setup

File: mobile/ios/Demo/Demo/AppDelegate.h [hunk 7]
```diff
@@ -0,0 +1,17 @@
+//
+//  AppDelegate.h
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Demo/Demo/AppDelegate.h]: Remove authors from comments please

File: mobile/ios/Demo/Demo/AppDelegate.m [hunk 8]
```diff
@@ -0,0 +1,51 @@
+//
+//  AppDelegate.m
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Demo/Demo/AppDelegate.m]: Remove

File: mobile/ios/Demo/Demo/Constants.h [hunk 9]
```diff
@@ -0,0 +1,16 @@
+//
+//  Constants.h
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Demo/Demo/Constants.h]: remove

File: mobile/ios/Demo/Demo/Constants.m [hunk 10]
```diff
@@ -0,0 +1,14 @@
+//
+//  Constants.m
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Demo/Demo/Constants.m]: remove

File: mobile/ios/Demo/Demo/ViewController.h [hunk 11]
```diff
@@ -0,0 +1,17 @@
+//
+//  ViewController.h
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Demo/Demo/ViewController.h]: remove

File: mobile/ios/Demo/Demo/ViewController.m [hunk 12]
```diff
@@ -0,0 +1,64 @@
+//
+//  ViewController.m
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Demo/Demo/ViewController.m]: remove

File: mobile/ios/Demo/Demo/main.m [hunk 13]
```diff
@@ -0,0 +1,16 @@
+//
+//  main.m
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Demo/Demo/main.m]: remove

File: mobile/ios/Demo/DemoTests/DemoTests.m [hunk 14]
```diff
@@ -0,0 +1,49 @@
+//
+//  EntangledDemoTests.m
+//  EntangledDemoTests
+//
+//  Created by Rajiv Shah on 6/26/18.
```
Comment (th0br0) [inline mobile/ios/Demo/DemoTests/DemoTests.m]: remove

File: mobile/ios/Demo/DemoTests/TestConstants.h [hunk 15]
```diff
@@ -0,0 +1,16 @@
+//
+//  TestConstants.h
+//  Demo
+//
+//  Created by Rajiv Shah on 6/26/18.
```
Comment (th0br0) [inline mobile/ios/Demo/DemoTests/TestConstants.h]: remove

File: mobile/ios/Demo/DemoTests/TestConstants.m [hunk 16]
```diff
@@ -0,0 +1,16 @@
+//
+//  TestConstants.m
+//  EntangledDemo
+//
+//  Created by Rajiv Shah on 6/26/18.
```
Comment (th0br0) [inline mobile/ios/Demo/DemoTests/TestConstants.m]: remove

File: mobile/ios/Interface.h [hunk 17]
```diff
@@ -0,0 +1,17 @@
+//
+//  Interface.h
+//  EntangledKit
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Interface.h]: remove

File: mobile/ios/Interface.h [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+//  Copyright © 2018 IOTA Foundation. All rights reserved.
+//
+
+
+#ifndef Interface_h
+#define Interface_h
```
Comment (th0br0) [inline mobile/ios/Interface.h]: please namespace the include guard

File: mobile/ios/Interface.m [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+//  Interface.m
+//  EntangledKit
+//
+//  Created by Rajiv Shah on 4/13/18.
```
Comment (th0br0) [inline mobile/ios/Interface.m]: remove

File: mobile/ios/Interface.h [hunk 20]
```diff
@@ -0,0 +1,17 @@
...
+
+char* doPOW(const char* trytes, int mwm);
```
Comment (th0br0) [inline mobile/ios/Interface.h]: we should also namespace these methods.

File: mobile/ios/Interface.m [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+//  Copyright © 2018 IOTA Foundation. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import "Interface.h"
+
+#include "common/helpers/sign.h"
+#include "common/helpers/pow.h"
+
+char* doPOW(const char* trytes, int mwm) {
```
Comment (th0br0) [inline mobile/ios/Interface.m]: Why do you even need the interface? the helpers from `common/helpers/*` are already providing these methods.

**Review by th0br0 — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 2 above)
Comment (rajivshah3) [reply .gitignore]: They were already there before, I only added from 75-146

File: mobile/ios/Demo/Demo/AppDelegate.h (hunk 7 above)
Comment (rajivshah3) [reply mobile/ios/Demo/Demo/AppDelegate.h]: Whoops, will do. I think Xcode automatically generates that

File: WORKSPACE (hunk 4 above)
Comment (rajivshah3) [reply WORKSPACE:40]: Resolved in https://github.com/iotaledger/entangled/pull/26/commits/84cebe5a093f98dfbf68040ed69e5ef06b2af9db

**Review by jserv — COMMENTED** [review]
I am a bit curious why filename existed as  
`mobile/ios/Demo/Demo/Constants.m` rather than `mobile/ios/Demo/Constants.m`?

**Comment by rajivshah3** [comment]
@th0br0 I was able to remove the `Interface.m` file and `objc_library` rule so that the build process is a bit more efficient, but I still can't get it to compile without `Interface.h`. I did namespace it though as requested.

@jserv I think it was so that I could keep all of the files for the demo project hidden in a single folder since they're not explicitly required for the framework

File: common/pow/pow.c [hunk 22]
```diff
@@ -8,6 +8,7 @@
 #include <stdlib.h>
 #include <string.h>
 
+
```
Comment (jserv) [inline common/pow/pow.c]: Is this change necessary?

Comment (rajivshah3) [reply common/pow/pow.c]: Nope, that must have been added when I was resolving merge conflicts. I fixed all unnecessary whitespace changes in https://github.com/iotaledger/entangled/pull/26/commits/dd9de0d6dab1058ccd64e07049b269dbc7bc2c60

**Comment by rajivshah3** [comment]
@th0br0 ready for another review when you get a chance

**Comment by thibault-martinez** [comment]
Hi @rajivshah3 , can you apply clang-format please ? Buildkite is complaining.

**Comment by rajivshah3** [comment]
@thibault-martinez done in 8ec0b60

**Comment by rajivshah3** [comment]
Awaiting iotaledger/ci-pipelines#19, once that's merged the tests should pass

**Review by thibault-martinez — APPROVED** [review]
